src/main/java/com/limechain/storage/block/BlockState.java

```diff
@@ -0,0 +1,798 @@
+            allDescendants = new ArrayList<>();
+            allDescendants.add(hash);
+        }
+
+        BlockHeader header = getHeader(hash);
+
+        List<byte[]> nextBlockHashes = getHashesByNumber(header.getBlockNumber().add(BigInteger.ONE));
+
+        for (byte[] nextBlockHash : nextBlockHashes) {
+            BlockHeader nextHeader = getHeader(nextBlockHash);
+
+            if (!Arrays.equals(nextHeader.getParentHash().getBytes(), hash)) {
+                continue;
+            }
+
+            List<byte[]> nextDescendants = getAllDescendants(nextBlockHash);
+            allDescendants.addAll(nextDescendants);
+        }
+
+        return allDescendants;
+    }
+
+    /**
+     * Get the block header on our best chain with the given number
+     *
+     * @param num the block number
+     * @return the block header
+     */
+    public BlockHeader getHeaderByNumber(BigInteger num) {
+        byte[] hash = getHashByNumber(num);
+        return getHeader(hash);
+    }
+
+    /**
+     * Get the block on our best chain with the given number
+     *
+     * @param num the block number
+     * @return the block
+     */
+    public Block getBlockByNumber(BigInteger num) {
+        byte[] hash = getHashByNumber(num);
+
+        return getBlockByHash(hash);
+    }
+
+    /**
+     * Get the block with the given hash
+     *
+     * @param hash the block hash
+     * @return the block
+     */
+    public Block getBlockByHash(byte[] hash) {
+        // Assuming unfinalizedBlocks is a map or similar structure with a getBlock method
+        Block block = unfinalizedBlocks.get(hash);
+        if (block != null) {
+            return block;
+        }
+
+        BlockHeader header = getHeader(hash);
+
+        BlockBody blockBody = getBlockBody(hash);
+        return new Block(header, blockBody);
+    }
+
+    /**
+     * Persist the block header in the database
+     *
+     * @param header the block header to be persisted
+     * @return true if the block header was successfully persisted, false otherwise
+     */
+    public boolean setHeader(final BlockHeader header) {
+        byte[] byteArray = helper.writeHeader(header);
+        return db.save(helper.headerKey(header.getHash()), byteArray);
+    }
+
+    /**
+     * Checks if the block body is in the database
+     *
+     * @param hash the block hash
+     * @return true if the block body is in the database, false otherwise
+     */
+    public boolean hasBlockBody(byte[] hash) {
+        if (unfinalizedBlocks.containsKey(hash)) {
+            return true;
+        }
+
+        return db.find(helper.blockBodyKey(hash)).isPresent();
+    }
+
+    /**
+     * Get the block body for a given hash
+     *
+     * @param hash the block hash
+     * @return the block body
+     */
+    public BlockBody getBlockBody(byte[] hash) {
+        // Assuming unfinalizedBlocks is a map or similar structure with a getBlockBody method
+        Block block = unfinalizedBlocks.get(hash);
```

Comment:
What does this comment mean? Are we not sure what the data structure of `unfinalizedBloks` is?

Reply:
It was added during the planning of the methods and forgotten after that